interface/mmal/vc/mmal_vc_client.c

```diff
@@ -82,6 +83,143 @@ struct MMAL_CLIENT_T
    MMAL_BOOL_T inited;
 };
 
+/*****************************************************************************
+ * Lookup table functions for client_component handles.
+ * Required as the IPC is strictly 32bit, therefore 64bit userland can not
+ * pass in the required pointers.
+ *****************************************************************************/
+#define MAX_COMPONENT_HANDLES 64
+
+typedef struct
+{
+   int index;
+   unsigned inuse;
+   MMAL_COMPONENT_T *component;
+} MMAL_CLIENT_COMPONENT_T;
+
+typedef struct
+{
+   MMAL_CLIENT_COMPONENT_T components[MAX_COMPONENT_HANDLES];
+   VCOS_MUTEX_T lock;
+} MMAL_CLIENT_COMPONENT_POOL_T;
+
+static MMAL_CLIENT_COMPONENT_POOL_T client_component_pool;
+
+uint32_t mmal_vc_allocate_client_component(MMAL_COMPONENT_T *component)
+{
+   int i;
+
+   vcos_mutex_lock(&client_component_pool.lock);
+   for (i=0; i<MAX_COMPONENT_HANDLES; i++)
+   {
+      if (client_component_pool.components[i].inuse == 0)
+         break;
+   }
+
+   if (vcos_verify(i != MAX_COMPONENT_HANDLES))
+   {
+      client_component_pool.components[i].index = i;
+      client_component_pool.components[i].component = component;
+      client_component_pool.components[i].inuse = 1;
+   }
+   vcos_mutex_unlock(&client_component_pool.lock);
+
+   return i;
+}
+
+static MMAL_COMPONENT_T *lookup_client_component(int index)
+{
+   if (vcos_verify(index < MAX_COMPONENT_HANDLES))
+   {
+      vcos_assert(client_component_pool.components[index].inuse);
+      return client_component_pool.components[index].component;
+   }
+
+   return NULL;
+}
+
+void mmal_vc_release_client_component(MMAL_COMPONENT_T *component)
+{
+   int i;
+
+   vcos_mutex_lock(&client_component_pool.lock);
+   for (i=0; i<MAX_COMPONENT_HANDLES; i++)
+   {
+      if (client_component_pool.components[i].component == component)
+      {
+         client_component_pool.components[i].component = NULL;
+         client_component_pool.components[i].inuse = 0;
+      }
+   }
+   vcos_mutex_unlock(&client_component_pool.lock);
+}
+
+#define MAX_CLIENT_CONTEXTS 128
+
+typedef struct
+{
+   int index;
+   unsigned inuse;
+   MMAL_VC_CLIENT_BUFFER_CONTEXT_T *ctx;
+} MMAL_CLIENT_CONTEXT_T;
+
+typedef struct
+{
+   MMAL_CLIENT_CONTEXT_T contexts[MAX_CLIENT_CONTEXTS];
+   VCOS_MUTEX_T lock;
+} MMAL_CLIENT_CONTEXT_POOL_T;
+
+static MMAL_CLIENT_CONTEXT_POOL_T client_context_pool;
+
+uint32_t mmal_vc_allocate_client_context(MMAL_VC_CLIENT_BUFFER_CONTEXT_T *context)
+{
+   int i;
+
+   vcos_mutex_lock(&client_context_pool.lock);
+   for (i=0; i<MAX_CLIENT_CONTEXTS; i++)
+   {
+      if (client_context_pool.contexts[i].inuse == 0)
+         break;
+   }
+
+   if (vcos_verify(i != MAX_CLIENT_CONTEXTS))
+   {
+      client_context_pool.contexts[i].index = i;
+      client_context_pool.contexts[i].ctx = context;
+      client_context_pool.contexts[i].inuse = 1;
+   }
+   vcos_mutex_unlock(&client_context_pool.lock);
+
+   return i;
+}
```

Comment:
Has this been tested when MAX_CLIENT_CONTEXTS has been reached?
It looks like this returns MAX_CLIENT_CONTEXTS, so mmal_vc_lookup_client_context returns NULL.
But that doesn't seem to be protected against (e.g. mmal_vc_port_by_number dereferences null pointer)

Reply:
No, the failure mode is currently going to be nasty. The limits are set so that they are highly unlikely to ever be hit, but the alloc does need to fail gracefully and then pass the failure back up.